server/app/repository/VersionRepository.java

```diff
@@ -417,6 +424,24 @@ public boolean isActiveProgram(Long programId) {
         .anyMatch(activeProgram -> activeProgram.id.equals(programId));
   }
 
+  /** Validate all programs have associated questions. */
+  private void validateProgramQuestionState() {
+    Version activeVersion = getActiveVersion();
+    Set<Long> newActiveQuestionIds =
+        activeVersion.getQuestions().stream()
+            .map(question -> question.getQuestionDefinition().getId())
+            .collect(Collectors.toSet());
+    Set<Long> missingQuestions =
+        activeVersion.getPrograms().stream()
+            .map(program -> getQuestionsInProgram(program.getProgramDefinition()))
+            .flatMap(Collection::stream)
+            .filter(question -> !newActiveQuestionIds.contains(question))
+            .collect(Collectors.toSet());
+    if (!missingQuestions.isEmpty()) {
+      throw new BadRequestException(String.format("Questions not found: %s", missingQuestions));
```

Comment:
**suggestion (blocking):** 

1. If we hit this case it's not the result of a bad request but because of a bug in the server code, so the exception here needs to produce a 500 rather than a 400.

2. Please include more info in the exception message to aid debugging:

```suggestion
      throw new IllegalStateException(String.format("Illegal state encountered when attempting to publish a new version. Question IDs found in program definitions %s not found in new active version: %s", programsIdsWithMissingQuestions, missingQuestionIdss));
```

Reply:
Yeah good call - updated to something similar.